src/index.js

```diff
@@ -37,6 +37,22 @@ function cachedIsFile (file, cb) {
 	isFileCache[file].then(contents => cb(null, contents), cb);
 }
 
+let isDirCache = {};
+function cachedIsDir (dir, cb) {
+	if (dir in isDirCache === false) {
+		isDirCache[dir] = statAsync(dir)
+			.then(
+				stat => stat.isDirectory(),
+				err => {
+					if (err.code === 'ENOENT') return false;
+					delete isDirCache[dir];
+					throw err;
+				});
+	}
+	isDirCache[dir].then(contents => cb(null, contents), cb);
+}
```

Comment:
I wonder how much overlap there is between the arguments supplied to `cachedIsDir` and `cachedIsFile`.  Maybe it would provide even better performance to roll both `cachedIsDir` and `cachedIsFile` into one function `cachedStat`, or maybe add an intermediate caching layer for caching `stat` and then have `cachedIsDir` and `cachedIsFile` just use `cachedStat`. Would you be willing to do some experimenting/benchmarking with this? Of course if there is hardly any overlap between the arguments to these functions, it will not be very useful.

Reply:
The numbers show its marginally slower to implement something like `cachedStat` for most cases; I have a feeling this is because the `resolve` algo is already good at determining what is a directory and what is a file, so it tends not to repeat stat calls between files/directories. At the moment the only callsite for `isDirectory` is effectively `path.basename(importee)` and if https://github.com/browserify/resolve/pull/190 gets merged that just expands to checking for `node_modules` also.